File: java/com/joshmanisdabomb/lcc/gui/inventory/AtomicBombScreen.java
package com.joshmanisdabomb.lcc.gui.inventory;

import com.joshmanisdabomb.lcc.LCC;
import com.joshmanisdabomb.lcc.container.AtomicBombContainer;
import com.joshmanisdabomb.lcc.network.AtomicBombDetonatePacket;
import com.joshmanisdabomb.lcc.network.LCCPacketHandler;
import com.mojang.blaze3d.platform.GlStateManager;
import net.minecraft.client.gui.screen.inventory.ContainerScreen;
import net.minecraft.client.gui.widget.Widget;
import net.minecraft.client.resources.I18n;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.fml.network.PacketDistributor;

import java.awt.*;
import java.util.function.BiConsumer;
import java.util.function.IntSupplier;

public class AtomicBombScreen extends ContainerScreen<AtomicBombContainer> {

    private static final ResourceLocation GUI = new ResourceLocation(LCC.MODID, "textures/gui/container/atomic_bomb.png");

    private FunctionalSpriteButton buttonDetonate;

    public AtomicBombScreen(AtomicBombContainer container, PlayerInventory playerInv, ITextComponent textComponent) {
        super(container, playerInv, textComponent);
        this.xSize = 176;
        this.ySize = 171;
    }

    @Override
    protected void init() {
        super.init();

        this.buttonDetonate = this.addButton(new DetonateButton(() -> {
            LCCPacketHandler.send(PacketDistributor.SERVER.noArg(), new AtomicBombDetonatePacket(container.te.getWorld().getDimension().getType(), container.te.getPos(), playerInventory.player.getUniqueID()));
            minecraft.player.closeScreen();
            return -1;
        }, (x, y) -> this.renderTooltip(I18n.format("gui.lcc.atomic_bomb.detonate"), x, y), GUI, this.guiLeft + 77, this.guiTop + 47, 88, 0, 171));
        this.buttonDetonate.active = container.te.canDetonate();
    }

    @Override
    public void tick() {
        this.buttonDetonate.active = container.te.canDetonate();
    }

    @Override
    public void render(int mouseX, int mouseY, float partialTicks) {
        this.renderBackground();
        super.render(mouseX, mouseY, partialTicks);
        this.renderHoveredToolTip(mouseX, mouseY);
        for (Widget widget : this.buttons) {
            if (widget.isHovered()) {
                widget.renderToolTip(mouseX, mouseY);
                break;
            }
        }
    }

    @Override
    protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
        this.font.drawString(this.title.getFormattedText(), 8.0F, 6.0F, 4210752);
        this.font.drawString(this.playerInventory.getDisplayName().getFormattedText(), 8.0F, (float)(this.ySize - 94), 4210752);
    }

    @Override
    protected void drawGuiContainerBackgroundLayer(float partialTicks, int mouseX, int mouseY) {
        GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
        this.minecraft.getTextureManager().bindTexture(GUI);
        this.blit(this.guiLeft, this.guiTop, 0, 0, this.xSize, this.ySize);
    }

    private static class DetonateButton extends FunctionalSpriteButton {

        public DetonateButton(IntSupplier onPress, BiConsumer<Integer, Integer> tooltip, ResourceLocation texture, int x, int y, int ix, int sx, int sy) {
            super(onPress, tooltip, texture, x, y, ix, sx, sy);
        }

        @Override
        protected void onRenderButton() {
            if (this.active && this.isHovered()) {
                Color c = new Color(Color.HSBtoRGB((float)((Math.cos(((System.currentTimeMillis() % 2000) / 2000F) * Math.PI * 2) * 0.03F) + 0.11F), 0.8F, 1.0F));
                GlStateManager.color4f(c.getRed() / 255F, c.getGreen() / 255F, c.getBlue() / 255F, 1.0F);
            } else {
                super.onRenderButton();
            }
        }

    }
}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/gui/screen/AtomicBombScreen.kt
package com.joshmanisdabomb.lcc.gui.screen

import com.joshmanisdabomb.lcc.LCC
import com.joshmanisdabomb.lcc.directory.LCCPacketsToServer
import com.joshmanisdabomb.lcc.inventory.container.AtomicBombScreenHandler
import com.joshmanisdabomb.lcc.utils.FunctionalButtonWidget
import com.mojang.blaze3d.platform.GlStateManager
import com.mojang.blaze3d.systems.RenderSystem
import io.netty.buffer.Unpooled
import net.fabricmc.fabric.api.network.ClientSidePacketRegistry
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.ingame.HandledScreen
import net.minecraft.client.render.GameRenderer
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.Inventory
import net.minecraft.network.PacketByteBuf
import net.minecraft.text.Text
import net.minecraft.text.TranslatableText
import net.minecraft.util.math.MathHelper
import kotlin.math.cos

class AtomicBombScreen(handler: AtomicBombScreenHandler, inventory: PlayerInventory, title: Text) : HandledScreen<AtomicBombScreenHandler>(handler, inventory, title) {

    val listener = ::onChanged
    val detonate: FunctionalButtonWidget by lazy { DetonateButton(x + 77, y + 47) {
        run {
            val world = MinecraftClient.getInstance().world?.registryKey ?: return@run
            val pos = handler.pos ?: return@run
            ClientSidePacketRegistry.INSTANCE.sendToServer(LCCPacketsToServer[LCCPacketsToServer::atomic_bomb_detonate].first().id, PacketByteBuf(Unpooled.buffer()).apply { writeIdentifier(world.value); writeBlockPos(pos) })
        }
        this.onClose()
        return@DetonateButton null
    } }

    init {
        handler.inventory.addListener(listener)
    }

    override fun init() {
        super.init()
        backgroundWidth = 176
        backgroundHeight = 171
        playerInventoryTitleX = 8
        playerInventoryTitleY = 77
        addButton(detonate)
        onChanged(handler.inventory)
    }

    override fun resize(client: MinecraftClient, width: Int, height: Int) {
        super.resize(client, width, height)
        detonate.x = x + 77
        detonate.y = y + 47
    }

    override fun onClose() {
        handler.inventory.removeListener(listener)
        super.onClose()
    }

    override fun drawBackground(matrices: MatrixStack, delta: Float, mouseX: Int, mouseY: Int) {
        RenderSystem.setShader(GameRenderer::method_34542)
        RenderSystem.setShaderTexture(0, texture)
        RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F)
        drawTexture(matrices, x, y, 0, 0, backgroundWidth, backgroundHeight)
    }

    override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
        renderBackground(matrices)
        super.render(matrices, mouseX, mouseY, delta)
        drawMouseoverTooltip(matrices, mouseX, mouseY)
        for (widget in buttons) {
            if (widget.isHovered) widget.renderToolTip(matrices, mouseX, mouseY)
            break
        }
    }

    fun onChanged(inventory: Inventory) {
        detonate.active = handler.inventory.canDetonate
    }

    companion object {
        val texture = LCC.id("textures/gui/container/atomic_bomb.png")
    }

    private inner class DetonateButton(x: Int, y: Int, pressed: () -> Int?) : FunctionalButtonWidget(x, y, 22, 22, 22, 22, TranslatableText("gui.lcc.atomic_bomb.detonate"), { matrices, x, y -> this@AtomicBombScreen.renderOrderedTooltip(matrices, textRenderer.wrapLines(TranslatableText("gui.lcc.atomic_bomb.detonate"), Int.MAX_VALUE) + textRenderer.wrapLines(TranslatableText("gui.lcc.atomic_bomb.detonate.power", handler.inventory.uraniumCount), Int.MAX_VALUE), x, y) }, pressed) {

        init {
            texture = Companion.texture
            ix = width * 4
            sx = 0
            sy = backgroundHeight
        }

        override fun onRenderButton(matrices: MatrixStack) {
            if (!disabled && active && isHovered) {
                val color = MathHelper.hsvToRgb(cos(System.currentTimeMillis().rem(2000).div(2000F).times(Math.PI.toFloat()).times(2)).times(0.03F) + 0.11F, 0.8F, 1.0F)
                val red = (color shr 16 and 255).toFloat() / 255.0f
                val green = (color shr 8 and 255).toFloat() / 255.0f
                val blue = (color and 255).toFloat() / 255.0f
                RenderSystem.setShaderColor(red, green, blue, 1f)
            } else {
                super.onRenderButton(matrices)
            }
        }

    }

}